src/react/10-context-hooks.md

```diff
@@ -5,75 +5,224 @@
 
 @body
 
-# React Context via Hooks
+## The Prop Drilling Problem
 
-React's context api allows global app state to be shared among all of the components in a component tree. It's commonly used to store data that many components are interested in like the app's theme or the current user.
+A single component can take as many props as you want to give it, however just like arguments into functions it's a good idea to [limit this number](https://stackoverflow.com/questions/37695557/react-are-there-respectable-limits-to-number-of-props-on-react-components), as more props makes for a more confusing component.
 
-A context object can be created using `React.createContext()` which returns a `Provider` component that can wrap other components interested in the global state.
+This however, can be difficult to do when you have a lot of data to pass through your component tree. Consider the following hierarchy with a few "drilled props". Lets imagine that the `Theme`, `Domain` amd `RootUrl` are decided within the `App` component, but are __only__ needed within the `ButtonText` component. That is to say, `Dashboard` and `Button` have no business related to any of those props.
 
-```jsx title="Providing and Consuming a Theme with Context" subtitle="Use the Button with the Provider"
-import React, {createContext} from 'react';
+```js
+-- App(Theme, Domain, RootUrl)
+   |-- Dashboard(Theme, Domain, RootUrl)
+       |-- Button(Theme, Domain, RootUrl)
+           |-- ButtonText(Theme, Domain, RootUrl)
+       |-- Button(Theme, Domain, RootUrl)
+           |-- ButtonText(Theme, Domain, RootUrl)
+```
 
-const THEMES = {
-  blue: { color: '#0000ff', fontSize: '1.25rem' },
-  red: { color: '#7ea0ff', fontSize: '2rem' },
-};
+Without concerning ourselves with what exactly each of the props does, we can see the problem. Every piece of global state must be propagated through the entire hierarchy. As a result, each component would look something like this:
 
-export const ThemeContext = createContext()
+```js
+function Component1(props) {
+  return <Component2 Theme={props.Theme} Domain={props.Domain} RootUrl={props.RootUrl}>
+}
+```
 
-export default function Layout() {
+It is possible to re-write it using the spread operator, but what you gain in conciseness, you lose in clarity and performance. **Avoid writing components like this!**
+
+```js
+function Component1(props) {
+  return <Component2 {...props} >
+}
+```
+
+In this case, it doesn't matter if `Component1` even needs the `Theme` prop; it will always require it simply because `Component2` might require it. Many years ago, this was solved using libraries such as [Redux](https://redux.js.org/). These libraries would work by wrapping each component in a connector HoC (Higher-order Component) which would automatically pass in any required props.
```

Comment:
"Many years ago"
_Oof, I feel attacked ;)_

Reply:
I'm not sure how to deal with the mentions of external state management. It would be a disservice not the mention it, but also I feel like a minimalist slant would be good when the target are people who are learning React.

Another thing is that the Context API __was__ introduced as a native way to Redux-y stuff. Previously, it would've been absolutely necessary to stay sane on larger projects. Now its an extra. It's hard to present 2 options equally while the training has a clear bias towards simplicity.